### DISCUSSION TITLE
Entity Linker does not leverage custom candidate generation function during the training phase

### DISCUSSION THREAD

**Discussion by amirj** [explosion/spaCy#8161] [Help: Coding & Implementations] [answered]
I'm not sure if it's a bug in entity linker or not. I just defined a custom candidate generation function as follow:

```python
@registry.misc("scispacy_candidate_generator")
def create_candidates():
    """Custom candidate generation function"""
    return get_scispacy_candidates


def get_scispacy_candidates(kb, span):
    """Leverage SciSpaCy to generate a list of plausible candidates"""
    print("****")
    candidates = []
    ....
    return candidates
```

Add an entity_linker component to my pipeline as follow:

```python
config = {
                "labels_discard": [],
                "n_sents": 2,
                "incl_prior": True,
                "incl_context": True,
                "model": DEFAULT_NEL_MODEL,
                "entity_vector_length": self.entity_vector_length,
                "get_candidates": {"@misc": "scispacy_candidate_generator"},
}
self.nlp.add_pipe("entity_linker", config=config, last=True)
```

Train the model as follow:

```python
# initialise entity linker
el = self.nlp.get_pipe("entity_linker")
el.initialize(get_examples=self.get_examples, kb_loader=load_kb(self.loc_kb))
# train only the entity_linker
for epoch in range(epoches):
    with self.nlp.select_pipes(enable=["entity_linker"]):
        optimizer = self.nlp.resume_training()
        random.shuffle(self.train_set)
        # increasing batch sizes
        ds_train = tqdm(
            self.train_set, desc="learning...", total=len(self.train_set)
        )
        batches = minibatch(ds_train, size=compounding(4.0, 32.0, 1.001))
        losses = {}
        for batch in batches:
            self.nlp.update(batch, drop=0.1, losses=losses, sgd=optimizer)
        logger.info(f"epoch#{epoch+1} losses: {losses}")
```

It's strange that during the training phase, I can't see any instance of "***" indicator (first line in "get_scispacy_candidates" function), it means that the entity linker model does not leverage the custom candidate generation during the training phase at all; though It's being used during the inference, when calling the predict function.

Does it make sense?

**Comment by amirj** [comment]
To add more context around this issue, the following is the implementation of the function that generate the training set ("self.train_set" in the above code snippets):

```python
def get_examples(self, split: str = "trng"):
    """Get a doc bin from a dataset"""
    examples = []
    dataset = self.mm.export_to_structured_tuple(split=split)
    sentencizer = self.nlp.get_pipe("sentencizer")
    for text, annotation in dataset:
        try:
            doc = self.nlp.make_doc(text)
            example = Example.from_dict(doc, annotation)
            example.reference = sentencizer(example.reference)
            examples.append(example)
        except Exception as e:
            logger.warning(e)
    return examples
```

I checked the implementation of entity linker, but I couldn't find how the candidate generation function is called during the training phase. May I ask you to point me to the position where candidate generation function being called during the training please?

**Comment by svlandeg** [comment, marked as answer]
The current implementation of the Entity Linker is such that the candidate generation function is not a trainable component - it's just a function that will always return the same set of candidates, given the same input. The actual trainable component is the `entity_linker` itself, which needs to learn the similarities between the description of an entity in the knowledge base, and the sentences in which that same entity is being expressed. Currently, these similarities are learnt from positive examples only. 

All this means that the candidate generation step isn't actually needed at training time - the gold reference entities with their KB IDs are used as such.

**Comment by amirj** [comment]
Thanks @svlandeg for the clarification. Based on my experiments, this is the main reason that the entity linker component does not learn much after just one epoch, i.e. the model does not have enough chance to "make a prediction" and "learn from their mistakes". If you can provide more information about how to fix/improve that, I'm happy to contribute :-)

**Comment by amirj** [comment]
@svlandeg Is there any example of SpaCy models which are using both positive and negative examples? Something that I can leverage as a starting point!

Mt understanding is that the input sentences pass through [self.model.begin_update](https://github.com/explosion/spaCy/blob/202943bc8cc1491f05e8c8563d9b280b342dc827/spacy/pipeline/entity_linker.py#L252) to get the sentence_encodings, then leverage [get_loss function](https://github.com/explosion/spaCy/blob/202943bc8cc1491f05e8c8563d9b280b342dc827/spacy/pipeline/entity_linker.py#L262) to calculate the loss between the generated sentence_encodings and **gold** entity_encording. 

I can change the predict function to **predict** a kb_id (rather than using the gold kb_id); what I don't understand is that how can I ask the model to leverage this pair as a negative example?
